pdfs/src/Pdf/Scanner.hs

```diff
@@ -0,0 +1,66 @@
+
+module Pdf.Scanner where
+
+import Data.ByteString.Char8  as LBS
+import Data.ByteString.Base64 as LBS
+import Data.Hashable
+
+newtype PDFBody = PDFBody { getPDFBody :: LBS.ByteString }
+
+newtype MaxSearchLength = MaxSearchLength { getMaxSearchLength :: Maybe Int }
+
+inject :: MaxSearchLength -> LBS.ByteString -> PDFBody -> Maybe PDFBody
+inject
+    (fromMaybe maxBound . getMaxSearchLength -> quota)
+     thing
+    (PDFBody text)
+  =
+    let base64    = LBS.encode thing
+        commented = "\n%" <> fairCVStartMark <> "{" <> base64 <> "}%"
+    in  do
+        place <- findFromEnd quota insertionMark text
+        return $ PDFBody $ insertAt place commented text
+  where
+
+project :: MaxSearchLength -> PDFBody -> Maybe LBS.ByteString
+project
+    (fromMaybe maxBound . getMaxSearchLength -> quota)
+    (PDFBody text)
+  = do
+    place <- findFromEnd quota fairCVStartMark text
+    let after = LBS.drop (place + LBS.length fairCVStartMark + 1) text
+        piece = LBS.takeWhile (not . (== '}')) after
+
+    either (const Nothing) pure
+        $ LBS.decode piece
+
+findFromEnd :: Int -> LBS.ByteString -> LBS.ByteString -> Maybe Int
+findFromEnd quota what text = do
+    guard $ not $ LBS.null what
+    guard $ not $ LBS.null text
+    go quota (LBS.length text - LBS.length what)
+  where
+    go 0 _         = Nothing
+    go _ i | i < 0 = Nothing
+    go q i = do
+        if LBS.index text i == LBS.head what
+            && equalsFrom i what text
+        then Just i
+        else go (q - 1) (i - 1)
+
+equalsFrom :: Int -> LBS.ByteString -> LBS.ByteString -> Bool
+equalsFrom offset what text = do
+    what == LBS.take (LBS.length what) (LBS.drop offset text)
+
+insertAt :: Int -> LBS.ByteString -> LBS.ByteString -> LBS.ByteString
+insertAt place what text =
+    let (before, after) = LBS.splitAt place text
+    in   before <> what <> after
+
+insertionMark :: LBS.ByteString
+insertionMark = "\ntrailer"
```

Comment:
Funny, I opened my diplomas made in latex (ru and eng versions), and they both have no "trailer" section.

Reply:
Can you grep for it?
I definitely can't put stuff before `xref` section, b/c there is somewhere a "count of bytes from start to `xref`" field. I can put comment right after xref, though.

Reply:
I tried to grep and failed. That's weird though, any pdf format tutorial seems to say that the trailer section must be present.

Reply:
Its like `<html>` and `<body>` tags, I think. We can't really know how manglable PDF is, because its not the standard is used to show PDFs, but concrete PDF-clients.